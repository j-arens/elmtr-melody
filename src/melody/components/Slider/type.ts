import { DragProps, MelodyDragEvent } from '@components/DragHelper/type';

export type SliderOrientation = 'vertical' | 'horizontal';

export interface HandleOffsetStyle {
    transform: string;
    right: string;
}

export interface SliderBodyClickEvent {
    event: MouseEvent;
    offset: number;
    insideHandle: boolean;
}

export interface SliderDragEvent {
    event: MelodyDragEvent;
    offset: number;
}

export type SliderDragEventHandler = (data: SliderDragEvent) => any;

export interface SliderClasses {
    slider?: string;
    body?: string;
    backfill?: string;
    handle?: string;
}
