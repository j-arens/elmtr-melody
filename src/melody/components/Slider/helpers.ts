import { HandleOffsetStyle, SliderOrientation } from './type';

export const getHandlePlacement = (
    sliderArea: number,
    handleSize: number,
    offset: number,
): HandleOffsetStyle => {
    const placement = (sliderArea / handleSize) * offset;
    return {
        transform: `translate3d(${placement}%, 0, 0)`,
        right: `calc(100% - ${handleSize / 2}px)`,
    };
};

export const getNextOffset = (
    delta: number,
    base: number,
    area: number,
    orientation: SliderOrientation,
): number => {
    const offset = ((delta - base) / area) * 100;
    if (orientation === 'vertical') {
        return offset < 0 ? Math.abs(offset) : -Math.abs(offset);
    }
    return offset;
};
