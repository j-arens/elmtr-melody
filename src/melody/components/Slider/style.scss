.slider {
    display: flex;
    align-items: center;
    position: relative;
    width: 100%;

    &__body {
        position: relative;
        overflow: hidden;
        width: 100%;
        background-color: $color_mine_shaft;
    }

    &__backfill,
    &__handle {
        position: absolute;
        right: 100%;
    }

    &__backfill {
        width: 100%;
        height: 100%;
        background-color: $color_silver_chalice;
    }

    &__handle {
        cursor: pointer;
        border-radius: 50%;
        background-color: $color_white;
    }
}
