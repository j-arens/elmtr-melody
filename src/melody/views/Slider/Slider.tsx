import Slideshow from '@components/Artwork/Slideshow/';
import CatalystButton from '@components/controls/CatalystButton/';
import NextButton from '@components/controls/NextButton/';
import PrevButton from '@components/controls/PrevButton/';
import RepeatButton from '@components/controls/RepeatButton/';
import ShuffleButton from '@components/controls/ShuffleButton/';
import CurrentTrackArtist from '@components/CurrentTrackArtist/';
import CurrentTrackTitle from '@components/CurrentTrackTitle/';
import Dock from '@components/Dock';
import Glider from '@components/Glider/';
import CurrentTimeElapsed from '@components/time/CurrentTimeElapsed';
import CurrentTimeLeft from '@components/time/CurrentTimeLeft';
import VolumeCtrl from '@components/VolumeCtrl/';
import { prefixClasses } from '@utils/index';
import { h } from 'preact';
const s = require('./style.scss');

const volClasses = prefixClasses('vol-primary-color vol-hover-color vol-width');
const btnClasses = prefixClasses('btn-primary-color btn-hover-color btn-space');
const previewClasses = prefixClasses('preview-bg preview-min-height preview-padding');
const slideshowClasses = prefixClasses('slider-transition slider-timing');
const artworkClasses = prefixClasses('artwork-size artwork-repeat artwork-position artwork-attachment image-filters');
const titleClasses = prefixClasses('title-font title-shadow title-color');
const artistClasses = prefixClasses('artist-font artist-shadow artist-color');
const timeClasses = prefixClasses('time-font time-shadow time-color');

export default () => (
    <div class="melody__viewContainer" data-melody-view="slider">
        <div class={`${s.preview} ${previewClasses}`}>
            <Slideshow
                className={`${s.preview__slideshow} ${slideshowClasses}`}
                artworkClassName={artworkClasses}
            />
            <div class={s.preview__header}>
                <VolumeCtrl className={volClasses} />
                <Dock />
            </div>
            <div class={s.preview__marquee}>
                <CurrentTrackTitle className={`${s.preview__title} ${titleClasses}`} />
                <CurrentTrackArtist className={`${s.preview__artist} ${artistClasses}`} />
            </div>
            <div className={s.preview__glider}>
                <Glider />
            </div>
            <div class={s.preview__footer}>
                <CurrentTimeElapsed className={`${s.preview__time} ${timeClasses}`} />
                <CurrentTimeLeft className={`${s.preview__time} ${timeClasses}`} />
            </div>
        </div>
        <div class={`${s.controlbar} ${prefixClasses('controlbar')} melody-control-bar`}>
            <ShuffleButton className={btnClasses} />
            <PrevButton className={btnClasses} />
            <CatalystButton className={btnClasses} />
            <NextButton className={btnClasses} />
            <RepeatButton className={btnClasses} />
        </div>
    </div>
);
