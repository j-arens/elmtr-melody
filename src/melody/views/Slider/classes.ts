import { prefixClasses } from '@utils/index';
import { MELODY_CUSTOMIZATION_CLASS_PREFIX as c_prefix } from '../../constants';

export const controlBar = prefixClasses(c_prefix, 'controlbar');

export const volume = prefixClasses(
    c_prefix,
    'vol-primary-color',
    'vol-hover-color',
    'vol-width',
);

export const controls = prefixClasses(
    c_prefix,
    'btn-primary-color',
    'btn-hover-color',
    'btn-space',
);

export const preview = prefixClasses(
    c_prefix,
    'preview-bg',
    'preview-min-height',
    'preview-padding',
);

export const slideshow = prefixClasses(
    c_prefix,
    'slider-transition',
    'slider-timing',
);

export const artwork = prefixClasses(
    c_prefix,
    'artwork-size',
    'artwork-repeat',
    'artwork-position',
    'artwork-attachment',
    'image-filters',
);

export const title = prefixClasses(
    c_prefix,
    'title-font',
    'title-shadow',
    'title-color',
);

export const artist = prefixClasses(
    c_prefix,
    'artist-font',
    'artist-shadow',
    'artist-color',
);

export const time = prefixClasses(
    c_prefix,
    'time-font',
    'time-shadow',
    'time-color',
);

export const dockToggle = prefixClasses(
    c_prefix,
    'dock-toggle-color',
    'dock-toggle-size',
);
