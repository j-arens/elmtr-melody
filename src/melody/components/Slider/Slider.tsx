import { DragProps, MelodyDragEvent } from '@components/DragHelper/type';
import { ELEMENTOR_NO_DRAG } from '@constants';
import { GLOBAL } from '@melody/constants';
import { cySelector } from '@utils/index';
import * as throttle from 'lodash.throttle';
import { Component, h } from 'preact';
import { getHandlePlacement, getNextOffset } from './helpers';
import {
    SliderBodyClickEvent,
    SliderClasses,
    SliderDragEventHandler,
    SliderOrientation,
} from './type';
const s = require('./style.scss');

export interface StateProps {
    wrapperId: string;
}

interface HOCProps extends DragProps {
    onEditorChange: (cb: () => any) => void;
}

export interface OwnProps extends HOCProps {
    orientation?: SliderOrientation;
    bodySize: number;
    handleSize: number;
    offset: number;
    throttle?: number;
    eventRadius?: number;
    onHandleClick?: (e: MouseEvent) => any;
    onBodyClick?: (data: SliderBodyClickEvent) => any;
    onBeginDrag?: SliderDragEventHandler;
    onEndDrag?: SliderDragEventHandler;
    onDragging?: SliderDragEventHandler;
    classes?: SliderClasses;
}

interface State {
    area: number;
    base: number;
}

type Props = StateProps & OwnProps;

const defaultClasses = {
    slider: '',
    body: '',
    backfill: '',
    handle: '',
};

export default class extends Component<Props, State> {
    state = {
        area: 0,
        base: 0,
    };

    componentDidMount() {
        const { onEditorChange } = this.props;
        this.setDimensions();
        this.bindDragHandlers();
        onEditorChange(this.setDimensions);
        GLOBAL.addEventListener('resize', this.setDimensions);
    }

    componentWillUnmount() {
        GLOBAL.removeEventListener('resize', this.setDimensions);
    }

    bindDragHandlers() {
        const {
            onDragStart,
            onBeginDrag,
            onDrag,
            onDragging,
            onDragEnd,
            onEndDrag,
        } = this.props;
        onDragStart((e: MelodyDragEvent) => this.handleDragEvent(e, onBeginDrag));
        onDrag((e: MelodyDragEvent) => this.handleDragEvent(e, onDragging));
        onDragEnd((e: MelodyDragEvent) => this.handleDragEvent(e, onEndDrag));
    }

    setDimensions = throttle(() => {
        if (this.slider) {
            const { orientation } = this.props;
            const rect = this.slider.getBoundingClientRect();
            this.setState({
                area: orientation === 'vertical' ? rect.height : rect.width,
                base: orientation === 'vertical' ? rect.bottom : rect.left,
            });
        }
    }, 1000);

    setRef = (el?: HTMLElement) => {
        const { setDragRef } = this.props;
        if (el) {
            this.slider = el;
            setDragRef(el);
        }
    }

    handleDragEvent = throttle((
        event: MelodyDragEvent,
        handler?: SliderDragEventHandler | undefined,
    ) => {
        if (!this.slider || typeof handler !== 'function') {
            return;
        }
        const { orientation } = this.props;
        const { base, area } = this.state;
        const { clientX, clientY } = event;
        const delta = orientation === 'vertical' ? clientY : clientX;
        const nextOffset = getNextOffset(delta, base, area, orientation);
        handler({
            event,
            offset: nextOffset,
        });
    }, this.props.throttle = 0);

    onMouseDown = (event: MouseEvent) => {
        const { onBodyClick, orientation, offset, handleSize } = this.props;
        if (!this.slider || typeof onBodyClick !== 'function') {
            return;
        }
        const { base, area } = this.state;
        const { pageX, pageY } = event;
        const delta = orientation === 'vertical' ? pageY : pageX;
        const nextOffset = getNextOffset(delta, base, area, orientation);
        const overlap = (handleSize / this.state.area) * 100;
        const upperRange = offset + overlap;
        const lowerRange = offset - overlap;
        const insideHandle = (nextOffset >= lowerRange) && (nextOffset <= upperRange);
        onBodyClick({
            event,
            offset: nextOffset,
            insideHandle,
        });
    }

    slider: HTMLElement;

    render(props: Props, { area }: State) {
        const {
            offset,
            handleSize,
            eventRadius,
            onHandleClick,
            bodySize,
            classes,
        } = props;
        const userClasses = Object.assign(defaultClasses, classes);
        return (
            <div
                class={`${s.slider} ${ELEMENTOR_NO_DRAG} ${userClasses.slider}`}
                ref={this.setRef}
                onMouseDown={this.onMouseDown}
                style={{ padding: eventRadius ? `${eventRadius}px 0` : 0 }}
            >
                <div
                    class={`${s.slider__body} ${userClasses.body}`}
                    style={{ height: `${bodySize}px` }}
                    {...cySelector('slider-body')}
                >
                    <div
                        class={`${s.slider__backfill} ${userClasses.backfill}`}
                        style={{ transform: `translate3d(${offset}%, 0, 0)` }}
                        {...cySelector('slider-backfill')}
                    />
                </div>
                {area ?
                    <div
                        class={`${s.slider__handle} ${userClasses.handle}`}
                        onMouseDown={onHandleClick}
                        {...cySelector('slider-handle')}
                        style={{
                            ...getHandlePlacement(area, handleSize, offset),
                            width: `${handleSize}px`,
                            height: `${handleSize}px`,
                        }}
                    />
                    :
                    null
                }
            </div>
        );
    }
}
