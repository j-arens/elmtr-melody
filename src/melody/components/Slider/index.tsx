import { Component, h } from 'preact';
import { WithOptionalClassName } from '@components/type';
import { DragProps, MelodyDragEvent } from '@components/DragHelper/type';
import { GLOBAL } from '@melody/constants';
import { getHandlePlacement } from './helpers';
import { SliderColors, SliderBodyClickEvent, SliderDragEventHandler } from './type';
import DragHelper from '@components/DragHelper/';
import { ELEMENTOR_NO_DRAG } from '@constants';
const s = require('./style.scss');
const throttle = require('lodash.throttle');

// return (
//     <Slider
//         className={string}
//         height={string}
//         handleSize={string | number}
//         offset={number}
//         eventRadius={8}
//         onHandleClick={fn}
//         onBodyClick={fn}
//         onBeginDrag={fn}
//         onDragging={fn}
//         onEndDrag={fn}
//         colors={}
//     />
// );



interface Props extends WithOptionalClassName, DragProps {
    height: number;
    handleSize: number;
    offset: number;
    throttle: number;
    colors?: SliderColors;
    eventRadius?: number;
    onHandleClick?: (e: MouseEvent) => any;
    onBodyClick?: (data: SliderBodyClickEvent) => any;
    onBeginDrag?: SliderDragEventHandler
    onEndDrag?: SliderDragEventHandler
    onDragging?: SliderDragEventHandler
}

interface State {
    width: number;
    left: number;
}

class Slider extends Component<Props, State> {
    state = {
        width: 0,
        left: 0,
    }

    defaultColors = {
        bg: '#333',
        backfill: '#a5a5a5',
        handle: '#fff',
    }

    componentDidMount() {
        this.setDimensions();
        this.bindDragHandlers();
        GLOBAL.addEventListener('resize', this.setDimensions);
    }

    componentWillUnmount() {
        GLOBAL.removeEventListener('resize', this.setDimensions);
    }

    bindDragHandlers() {
        const {
            onDragStart,
            onBeginDrag,
            onDrag,
            onDragging,
            onDragEnd,
            onEndDrag,
        } = this.props;
        onDragStart((e: MelodyDragEvent) => this.handleDragEvent(e, onBeginDrag));
        onDrag((e: MelodyDragEvent) => this.handleDragEvent(e, onDragging));
        onDragEnd((e: MelodyDragEvent) => this.handleDragEvent(e, onEndDrag));
    }

    setDimensions = throttle(() => {
        if (this.slider) {
            const rect = this.slider.getBoundingClientRect();
            this.setState({
                width: rect.width,
                left: rect.left,
            });
        }
    }, 1000);

    setRef = (el?: HTMLElement) => {
        const { setDragRef } = this.props;
        if (el) {
            this.slider = el;
            setDragRef(el);
        }
    }

    getNextOffset(x: number): number {
        if (!this.slider) {
            return 0;
        }
        const { left, width } = this.state;
        return ((x - left) / width) * 100;
    }

    handleDragEvent = throttle((
        event: MelodyDragEvent,
        handler?: SliderDragEventHandler | undefined,
    ) => {
        if (!this.slider || typeof handler !== 'function') {
            return;
        }
        const { clientX } = event;
        const nextOffset = this.getNextOffset(clientX);
        handler({
            event,
            offset: nextOffset,
        });
    }, this.props.throttle = 0);

    onMouseDown = (event: MouseEvent) => {
        const { onBodyClick } = this.props;
        if (!this.slider || typeof onBodyClick !== 'function') {
            return;
        }
        const { pageX } = event;
        const { offset, handleSize } = this.props;
        const nextOffset = this.getNextOffset(pageX);
        const overlap = (handleSize / this.state.width) * 100
        const upperRange = offset + overlap;
        const lowerRange = offset - overlap;
        const insideHandle = (nextOffset >= lowerRange) && (nextOffset <= upperRange);
        onBodyClick({
            event,
            offset: nextOffset,
            insideHandle,
        });
    }

    slider: HTMLElement;

    render(props: Props, { width }: State) {
        const {
            offset,
            handleSize,
            eventRadius,
            onHandleClick,
            colors,
            height,
        } = props;
        const color = {
            ...this.defaultColors,
            ...colors,
        };
        return (
            <div
                class={`${s.slider} ${ELEMENTOR_NO_DRAG}`}
                ref={this.setRef}
                onMouseDown={this.onMouseDown}
                style={{
                    padding: eventRadius ? `${eventRadius}px 0` : 0, 
                }}
            >
                <div
                    class={s.slider__body}
                    style={{
                        height: `${height}px`,
                        backgroundColor: color.bg,
                    }}
                >
                    <div
                        class={s.slider__backfill}
                        style={{
                            transform: `translate3d(${offset}%, 0, 0)`,
                            backgroundColor: color.backfill,
                        }}
                    />
                </div>
                {width &&
                    <div
                        class={s.slider__handle}
                        onMouseDown={onHandleClick}
                        style={{
                            ...getHandlePlacement(width, handleSize, offset),
                            width: `${handleSize}px`,
                            height: `${handleSize}px`,
                            backgroundColor: color.handle,
                        }}
                    />
                }
            </div>
        );
    }
}

export default DragHelper(Slider);
