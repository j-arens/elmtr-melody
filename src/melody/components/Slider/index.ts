import { State } from '@redux/type';
import { connect } from 'preact-redux';
import Slider, { OwnProps, StateProps } from './Slider';

const mapState = (state: State) => ({
    wrapperId: state.ui.wrapperId,
});

let exportComponent = connect<StateProps, {}, OwnProps>(mapState)(Slider);

if (process.env.NODE_ENV === 'test') {
    exportComponent = Slider;
}

export default exportComponent;
