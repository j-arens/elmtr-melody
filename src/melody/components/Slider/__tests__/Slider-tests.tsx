import { h } from 'preact';
import { render } from 'preact-render-to-string';
import Slider from '../';

describe('Slider', () => {
    it('renders', () => {
        const component = render(
            <Slider
                height={5}
                handleSize={10}
                offset={0}
                eventRadius={10}
                classes={{
                    slider: 'lol-slider',
                    body: 'lol-body',
                    backfill: 'lol-backfill',
                    handle: 'lol-handle',
                }}
            />,
        );
        expect(component).toMatchSnapshot();
    });
});
