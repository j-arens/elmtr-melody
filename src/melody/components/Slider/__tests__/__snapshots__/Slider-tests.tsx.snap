// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders 1`] = `
<div class="slider elementor-tab-title lol-slider" style="padding: 10px 0;">
  <div class="slider__body lol-body" style="height: undefinedpx;" data-cy="slider-body">
    <div class="slider__backfill lol-backfill" style="transform: translate3d(0%, 0, 0);"
    data-cy="slider-backfill"></div>
  </div>
</div>
`;
